cpp/llm_chat.cc

```diff
@@ -931,61 +912,109 @@ class LLMChat {
     picojson::object config;
     config["temperature"] = picojson::value(this->temperature_);
     config["repetition_penalty"] = picojson::value(this->repetition_penalty_);
+    config["presence_penalty"] = picojson::value(this->presence_penalty_);
+    config["frequency_penalty"] = picojson::value(this->frequency_penalty_);
     config["top_p"] = picojson::value(this->top_p_);
     config["mean_gen_len"] = picojson::value(this->mean_gen_len_);
     config["max_gen_len"] = picojson::value(this->max_gen_len_);
     config["shift_fill_factor"] = picojson::value(this->shift_fill_factor_);
     config["conv_config"] = this->conversation_.SerializeToJSON();
     return picojson::value(config);
   }
-  /*!
-   * \brief Sample output token from logits on device
-   */
-  int32_t SampleTokenFromLogits(NDArray logits_on_device,
-                                picojson::object generation_config = picojson::object()) {
-    // prepare generation settings
-    // the generation_config will not override the original config
-    // since is only used for this generation
-    double gen_temperature;
-    double gen_repetition_penalty;
-    double gen_top_p;
+
+  picojson::object LoadGenerationConfigFromString(const std::string& generation_config_str) {
+    picojson::object generation_config = picojson::object();
+    if (!generation_config_str.empty()) {
+      picojson::value generation_config_json;
+      picojson::parse(generation_config_json, generation_config_str);
+      generation_config = generation_config_json.get<picojson::object>();
+    }
+    return generation_config;
+  }
+
+  void ReadGenerationConfig(picojson::object generation_config, double* gen_temperature,
```

Comment:
great PR @Kartik14!!
very small nit pick (not really an issue), isn't passing by ref this way, cleaner? all the functions follow this convention
```void ReadGenerationConfig(..., double& gen_temperature, NDArray& gen_temperature_arr, ...```
it's a little unnecessary adding * each time

Reply:
yeah, thanks for pointing it out, I will fix this in the next PR.